### PR TITLE
Allow running tests by line and column

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/test/issues/1579

- Allows querying tests by line & col using test path query strings
- Matches if any frame in the test stack trace matches up (this allows for lots of flexibility in terms of testing styles and parameterized tests).

~This does not currently work on browsers, but I probably could get it to? Not sure how important that ultimately is.~
- If a stack trace formatter is set up for the remote platform, it will use it to format the `trace` sent back for groups/tests that are discovered (enabling browser/node support). 

cc @DanTup

### CODE REVIEW

File: pkgs/test/README.md [hunk 1]
```diff
@@ -173,17 +173,31 @@ description (including any group descriptions) match that regular expression
...
 path/to/test.dart`.
 
+### Test Path Queries
+
+Some query parameters are supported on test paths, which allow you to filter the
+tests that will run within just those paths.
+
+- **name**: Works the same as `--name` (simple contains check).
+  - This is the only option that supports more than one entry.
+- **full-name**: Requires an match for the name of the test.
```
Comment (natebosch) [inline pkgs/test/README.md:190]:
```suggestion
- **full-name**: Requires an exact match for the name of the test.
```

File: pkgs/test/README.md [hunk 2]
```diff
@@ -173,17 +173,31 @@ description (including any group descriptions) match that regular expression
...
+- **line**: Matches any test that originates from this line in the test suite.
+  - Not supported for browser tests.
+- **col**: Matches any test that originates from this column in the test suite.
```
Comment (natebosch) [inline pkgs/test/README.md:193]: I'm confused who would use this? Is this a specific column? Like the position of the first `t` in `test(`?

File: pkgs/test_core/lib/src/runner.dart [hunk 3]
```diff
@@ -279,6 +280,14 @@ class Runner {
       }
     })).map((loadSuite) {
       return loadSuite.changeSuite((suite) {
+        // TODO: Support this for browser tests
+        if (suite.platform.runtime.isBrowser &&
```
Comment (natebosch) [inline pkgs/test_core/lib/src/runner.dart:284]: I think we don't support it for node tests either right?

It's probably more accurate to test `isJS` than `isBrowser`.

Comment (jakemac53) [reply pkgs/test_core/lib/src/runner.dart:284]: Good point

**Comment by jakemac53** [comment]
Added support and a test for browser/node tests

File: pkgs/test/test/runner/line_and_col_test.dart [hunk 4]
```diff
@@ -0,0 +1,361 @@
+// Copyright (c) 20121, the Dart project authors.  Please see the AUTHORS file
```
Comment (natebosch) [inline pkgs/test/test/runner/line_and_col_test.dart:1]:
```suggestion
// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
```

File: pkgs/test/README.md (hunk 2 above)
Comment (natebosch) [reply pkgs/test/README.md:192]: I still don't understand the precise position this matches.

_Is_ it exactly the first `t` in `test`? Is it 0 based? Does this match the same frame as the line number, or can the line match one frame and the column match a different frame?

I'll read through the test cases tomorrow to find out more, but I do think we should expand the user facing docs.

File: pkgs/test_core/lib/src/runner.dart [hunk 5]
```diff
@@ -298,6 +299,30 @@ class Runner {
             return false;
           }
 
+          // Skip tests that don't start on `line` or `col` if specified.
+          var line = suite.config.line;
+          var col = suite.config.col;
+          var trace = test.trace;
+          if (trace == null && (line != null || col != null)) {
```
Comment (natebosch) [inline pkgs/test_core/lib/src/runner.dart:306]: When would we be missing a trace? Should we consider it an error to try to run tests with a line and column when we have no trace available?

File: pkgs/test_core/lib/src/runner.dart [hunk 6]
```diff
@@ -298,6 +299,30 @@ class Runner {
...
+            var absoluteSuitePath = p.absolute(suite.path!);
+            if (line != null &&
+                !trace.frames.any((frame) =>
+                    frame.uri.scheme == 'file' &&
+                    frame.uri.toFilePath() == absoluteSuitePath &&
+                    frame.line == line)) {
+              return false;
+            }
+            if (col != null &&
+                !trace.frames.any((frame) =>
```
Comment (natebosch) [inline pkgs/test_core/lib/src/runner.dart:318]: Synced offline about enforcing line and column match the same frame, not any frame at all.

File: pkgs/test_core/lib/src/runner.dart (hunk 5 above)
Comment (jakemac53) [reply pkgs/test_core/lib/src/runner.dart:306]: I am not sure if we get load tests and things through this api. I will try making it an error and see what happens.

File: pkgs/test/CHANGELOG.md [hunk 7]
```diff
@@ -1,9 +1,13 @@
...
 
-* Support query parameters `name` and `full-name` on test paths, which will
-  apply the filters to only those test suites.
+* Support query parameters `name`, `full-name`, `line`, and `col` on test paths,
+  which will apply the filters to only those test suites.
   * All specified filters must match for a test to run.
   * Global filters (ie: `--name`) are also still respected and must match.
+  * The `line` and `col` will match if any frame from the test trace matches
+    (the test trace is the current stack trace where `test` is invoked).
+    * If either `line` or `col` are used, stack trace chaining is enabled for
```
Comment (natebosch) [inline pkgs/test/CHANGELOG.md:9]: Does this mainly matter for web tests? I can't find where in the code this happens - can you point me to it?

Comment (jakemac53) [reply pkgs/test/CHANGELOG.md:9]: It happens for all tests that run remotely (so for everything but the direct dart file invocation case), specifically here https://github.com/dart-lang/test/pull/1604/files#diff-d694d0b45775f9db69469d5db79e8b6f0bbb05b8ddcf110340a1e205c454f5a6R190

Comment (natebosch) [reply pkgs/test/CHANGELOG.md:9]: Ah, that's a different behavior. That isn't related to `--chain-stack-traces` or `Chain.capture`.

I think we can skip this mention from the user facing changelog - it's a minor detail of the communication between the runner and the tests, but it doesn't impact user visible behavior and shouldn't impact perf as much as stack trace chaining would.

Comment (jakemac53) [reply pkgs/test/CHANGELOG.md:9]: Ah ok, ya I was a bit confused by it

**Comment by jakemac53** [comment]
Fixes https://github.com/dart-lang/test/issues/1579

- Allows querying tests by line & col using test path query strings
- Matches if any frame in the test stack trace matches up (this allows for lots of flexibility in terms of testing styles and parameterized tests).
- If a stack trace formatter is set up for the remote platform, it will use it to format the `trace` sent back for groups/tests that are discovered (enabling browser/node support).

**Pull request merged**